src/docfx/lib/git/GitCommitProvider.cs

```diff
@@ -1,412 +1,83 @@
-                        singleParent = true;
-                        break;
-                    }
-                }
-
-                if (!singleParent)
-                {
-                    // Otherwise, follow all parents.
-                    for (var i = 0; i < parentCount; i++)
-                    {
-                        add = true;
-                        commitsToFollow.Add((commit.Parents[i], parentBlobs[i]));
-                    }
-                }
-
-                if (add)
-                {
-                    result.Add(commit);
-                }
-            }
-
-            if (updateCache)
+            foreach (var value in _commitProviders.Values)
             {
-                lock (commitCache)
+                if (value.IsValueCreated)
                 {
-                    _cacheUpdated = true;
-                    commitCache.Add((headCommit.Sha.a, headBlob), (result.Select(c => c.Sha.a).ToArray(), 0));
+                    var (provider, repo) = await value.Value;
+                    await GitCommitCacheProvider.SaveCache(repo, provider.BuildCache());
                 }
             }
-
-            return result.Select(c => c.GitCommit).ToList();
-        }
-
-        public Task SaveCache()
-        {
-            if (!_cacheUpdated || string.IsNullOrEmpty(_cacheFilePath))
-            {
-                return Task.CompletedTask;
-            }
-
-            PathUtility.CreateDirectoryFromFilePath(_cacheFilePath);
-
-            return ProcessUtility.WriteFile(_cacheFilePath, stream =>
-            {
-                using (var writer = new BinaryWriter(stream))
-                {
-                    // Create a snapshot of commit cache to ensure count and items matches.
-                    var commitCache = _commitCache.ToList();
-
-                    writer.Write(commitCache.Count);
-                    foreach (var (file, value) in commitCache)
-                    {
-                        lock (value)
-                        {
-                            writer.Write(file);
-                            writer.Write(Math.Min(value.Count, MaxCommitCacheCountPerFile));
-
-                            var lruValues = value.OrderBy(pair => pair.Value.lruOrder).Take(MaxCommitCacheCountPerFile);
-
-                            foreach (var ((commit, blob), (commitHistory, _)) in lruValues)
-                            {
-                                writer.Write(commit);
-                                writer.Write(blob);
-                                writer.Write(commitHistory.Length);
-
-                                foreach (var sha in commitHistory)
-                                {
-                                    writer.Write(sha);
-                                }
-                            }
-                        }
-                    }
-                }
-            });
         }
 
         public void Dispose()
```

Comment:
Do we still need the `Dispose` method? Since we have implemented the finalizer for `FileCommitProvider`, which will be invoked by GC to free unmanaged resources.

Reply:
the finalizer is just making sure it finally release the resource